api/python/quilt3/data_transfer.py

```diff
@@ -295,20 +302,50 @@ def _download_file(ctx, src_bucket, src_key, src_version, dest_path):
 
     dest_file.parent.mkdir(parents=True, exist_ok=True)
 
+    with dest_file.open('wb', buffering=0) as f:
+        fileno = f.fileno()
+        is_regular_file = stat.S_ISREG(os.stat(fileno).st_mode)
+
+        if is_regular_file:
+            # Preallocate file.
+            if hasattr(os, 'posix_fallocate'):
+                os.posix_fallocate(fileno, 0, size)
+            else:
+                f.truncate(size)
+
     if src_version is not None:
         params.update(dict(VersionId=src_version))
-    resp = s3_client.get_object(**params)
 
-    body = resp['Body']
-    with open(dest_path, 'wb') as fd:
-        while True:
-            chunk = body.read(64 * 1024)
-            if not chunk:
-                break
-            fd.write(chunk)
-            ctx.progress(len(chunk))
+    chunk_ranges = (
+        get_chunk_ranges(size)
+        if is_regular_file and size >= s3_transfer_config.multipart_threshold and os.name != 'nt' else
+        (None,)
+    )
+    remaining_counter = reversed(range(len(chunk_ranges)))
+
+    def download_part(chunk_range):
+        with dest_file.open('w+b') as chunk_f:
+            if chunk_range:
+                start, end = chunk_range
+                part_params = dict(params, Range=f'bytes={start}-{end}')
+                chunk_f.seek(start)
+            else:
+                part_params = params
+
+            resp = s3_client.get_object(**part_params)
+            body = resp['Body']
+            chunk_size = 64 * 1024
+            while True:
+                chunk = body.read(chunk_size)
+                if not chunk:
+                    break
+                ctx.progress(chunk_f.write(chunk))
 
-    ctx.done(PhysicalKey.from_path(dest_path))
+        if not next(remaining_counter):
```

Comment:
That's kind of a strange way to count... Just use a variable? Also, need a lock. See the other upload/download methods; something like:
```
            with lock:
                remaining -= 1
                done = remaining == 0
```

Reply:
> That's kind of a strange way to count.

I'm pretty sure it's atomic.

Reply:
Hm... Is it guaranteed? The only thing I can find is: if it's implemented in C, then it'll hold the GIL, so it will be atomic - and I'm guessing `next` on a `reversed`, `range`, etc. will be a C function. But, it looks like an implementation detail.

In any case, I think the cost of acquiring a lock is pretty minimal compared to the download of the part - so it's not worth risking data corruption even if we're 99% sure it's ok. (And who knows if someone decides to run this code using Jython or whatever else...)